doc: add tutorial for MutatingAdmissionPolicy using CEL

Champbreed opened the pull request:
This PR adds a new tutorial to the documentation covering MutatingAdmissionPolicy, a feature introduced in Alpha in Kubernetes v1.32. The tutorial provides a hands-on guide for users to understand how to use Common Expression Language (CEL) for resource mutation without writing a custom webhook.

Key sections added:
* Prerequisites: Setting up a local cluster (e.g., kind) with the necessary feature gates and API configurations.

* Policy Definition: Creating a MutatingAdmissionPolicy to automatically add labels to Pods.

* Policy Binding: Activating the policy using MutatingAdmissionPolicyBinding.

* Verification: Step-by-step instructions to verify that the mutation is applied successfully.

Verification Stage:
I have verified the YAML manifests and commands in this tutorial on a local kind cluster running Kubernetes v1.32.0.

* Confirmed MutatingAdmissionPolicy creates successfully with reinvocationPolicy: Never.

* Confirmed MutatingAdmissionPolicyBinding works correctly in v1alpha1.

* Verified that created Pods correctly receive the owner=champbreed label via the admission controller.

**AI Disclosure:** AI assistance was utilized specifically to help locate and format the correct `kubectl delete` commands utilized within the cleanup section of this tutorial.

Champbreed committed:
docs: add unified tutorial for Validating and Mutating Admission Policy

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,509 @@
+- liggitt
+- jpbetz
+- cici37
+- sftim
```
```suggestion
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,509 @@
+---
+<!-- overview -->
+
+{{< feature-state state="stable" for_k8s_version="v1.30" >}}
```
Leave these out; definitely don't use two.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,509 @@
+This page provides an overview of declarative admission policies, which allow you to use the Common Expression Language (CEL) to validate or mutate resources.
+
+* Mutating Admission Policy (Alpha): Modifies an object before it is stored (e.g., adding a default label).
+* Validating Admission Policy: Decides whether to allow or deny a request based on specific rules.
```
```suggestion
ValidatingAdmissionPolicy
: Configures whether to allow or deny a request that would modify a resource, based on specific rules.
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,509 @@
+
+<!-- body -->
+
+## Before you begin
```
Have a look at how other tutorials achieve this; they use a particular shortcode.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,509 @@
+You need to have a Kubernetes cluster, and the kubectl command-line tool must be configured to communicate with your cluster.
+
+* For **ValidatingAdmissionPolicy**, ensure your cluster is version 1.30 or later.
+* For **MutatingAdmissionPolicy**, you need:
```
(nit) avoid bold here

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,509 @@
+    * The `MutatingAdmissionPolicy` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) enabled.
+    * The `admissionregistration.k8s.io/v1alpha1` [API group](/docs/concepts/overview/kubernetes-api/#api-groups) enabled.
+
+### Local testing with kind
```
nit: could provide an explanation for kind and a different one for mknikube, using tabs to let people choose

lmktfy reviewed (commented):
Thanks for this! 

In this review I am providing feedback just on the start of the page; I did see that there are a few places we could tweak it.

Because tutorials often cover a lot and because the people reading them are especially likely to be new to learning Kubernetes, we like to have tutorials largely follow our [style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

---

The other important thing is testing that it all works. Briefly (and I mean that): how have you checked that this all works?

Champbreed commented:
Hi @lmktfy
I have addressed the pending feedbacks and also updated all Mutating Admission Policy references from v1alpha1 to v1beta1 to align with the K8s 1.32 release, corrected the labels from Alpha to Beta. Ready for another look!

Champbreed commented:
Hi @lmktfy, I’ve addressed all feedback:

Style Guide: Fixed overview description lists and removed bolding.

Feature State: Added v1.32 Beta shortcode to the overview.

API Updates: Updated all Mutating Admission Policy examples to v1beta1.

Git: Squashed into a single clean commit.

Ready for another look!

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,697 @@
+---
+<!-- overview -->
+
+{{< feature-state state="beta" for_k8s_version="v1.32" >}}
```
This is stale, I think.

Champbreed commented:
Hi @lmktfy, I’ve addressed the final feedback:

Removed the stale v1.32 banner (confirmed redundant vs peer docs).

Aligned styling (backticks for resource names) and verified the v1beta1 API versions.

Tested locally on a kind v1.32 cluster to ensure the tutorial is functional.

Commit squashed and force-pushed. Ready for another look! /lgtm

Champbreed commented:
/assign @tengqm

tengqm commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
````diff
@@ -0,0 +1,695 @@
+    - expression: "object.spec.replicas <= params.maxReplicas"
+      messageExpression: "'object.spec.replicas must be no greater than ' + string(params.maxReplicas)"
+```
+The spec.paramKind field specifies the kind of resources used to parameterize the policy. In this example, it is configured by ReplicaLimit custom resources. Note how the CEL expression references the parameters via the CEL params variable (e.g., params.maxReplicas). The spec.matchConstraints specifies what resources this policy is designed to evaluate. Native types such as ConfigMap can also be  ...
````
```suggestion
The `spec.paramKind` field specifies the kind of resources used to parameterize the policy.
In this example, it is configured by ReplicaLimit custom resources.
Note how the CEL expression references the parameters via the CEL `params` variable (e.g., params.maxReplicas).
The `spec.matchConstraints` specifies what resources this policy is designed to evaluate.
Native types such as ConfigMap can also be used as parameter references.
```

tengqm commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
````diff
@@ -0,0 +1,695 @@
+  name: "replica-limit-prod.example.com"
+maxReplicas: 100
+```
+For each admission request, the API server evaluates CEL expressions of each (policy, binding, param) combination that match the request. For a request to be admitted, it must pass all evaluations.
````
When referencing to field/property names, it is better to use backticks for highlighting.

```suggestion
For each admission request, the API server evaluates CEL expressions of each (`policy`, `binding`, `param`) combination that match the request.
For a request to be admitted, it must pass all evaluations.
```

lmktfy commented:
> Semantic Line Breaks

I don't know that standard. Could you summarize it in 200 words of pentameter?

Champbreed committed:
Remove reviewers from admission policy tutorial

Removed reviewers from the admission policy tutorial.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,292 @@
+Each admission policy binding must specify one or more actions
+to declare how the policy is enforced.
+
+For `ValidatingAdmissionPolicyBinding`,
```
```suggestion
For ValidatingAdmissionPolicyBinding,
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,292 @@
+  [warning](/blog/2020/09/03/warnings/).
+
+`Audit`
+: Validation failure is included in the audit event for the API request.
```
(nit)
I suggest either alphabetical order, or Audit &rarr; Warn &rarr; Deny (putting them in in order of increasing visibility to the user).

lmktfy reviewed (commented):
Thanks for the PR. With some more work - possibly more than two hours - we could surely merge it.

However if this were merged as-is, I would be calling for us (SIG Docs) to either quickly improve it, or revert. It's that far from the typical quality we like our docs to meet.

Champbreed committed:
docs: address reviewer feedback on MutatingAdmissionPolicy tutorial

Champbreed committed:
docs: align cleanup section with tutorial resources

lmktfy commented:
Ok, noted. Would you be willing to update the PR description to make that clear?

That's the text that, as I write this, begins _“This PR adds a new tutorial to the documentation covering MutatingAdmissionPolicy, a feature introduced in Alpha in Kubernetes v1.32…”_

Champbreed commented:
Ready for your review. @lmktfy

tengqm reviewed (commented):
/lgtm

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
````diff
@@ -0,0 +1,303 @@
+
+For MutatingAdmissionPolicy, you need:
+* A cluster running version 1.32 or later.
+* The `MutatingAdmissionPolicy` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) enabled.
+* The `admissionregistration.k8s.io/v1beta1` [API group](/docs/concepts/overview/kubernetes-api/#api-groups) enabled.
+
+### Local testing
+
+{{< tabs name="local_cluster_setup" >}}
+{{% tab name="kind" %}}
+If you are using [kind](https://kind.sigs.k8s.io/), 
+use this configuration to enable the necessary features:
+
+```yaml
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+featureGates:
+  "MutatingAdmissionPolicy": true
+runtimeConfig:
+  "admissionregistration.k8s.io/v1beta1": "true"
+nodes:
+- role: control-plane
+  image: kindest/node:v1.32.0
+```
+
+{{% /tab %}} 
+{{% tab name="minikube" %}} 
+If you are using minikube, 
+run:
+
+```bash
+minikube start --feature-gates=MutatingAdmissionPolicy=true \
+  --runtime-config=admissionregistration.k8s.io/v1beta1=true
+```
+
+{{% /tab %}} 
+{{< /tabs >}}
````
```suggestion
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+A declarative policy requires a Policy and a Binding. 
+Parameter resources are optional and provide runtime configuration.
+
+The `policy` resource
```
No need for backticks here.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+  while a MutatingAdmissionPolicy can modify resources 
+  such as adding a default label to a namespace.
+
+The `binding` resource
```
No need for backticks here.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+  If you only want to enforce a policy for a specific subset of resources, 
+  the binding is where you narrow the scope of the policy using `matchResources`.
+
+The `parameter` resource (optional)
```
No need for backticks here.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+The `parameter` resource (optional)
+: Allows a policy configuration to be separate from its definition. 
+  Parameter resources refer to Kubernetes resources available in the API. 
+  They can be built-in types like `ConfigMap` 
```
No need for backticks here.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+
+### ValidatingAdmissionPolicy
+
+The following is an example of a `ValidatingAdmissionPolicy`
```
No need for backticks here.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
````diff
@@ -0,0 +1,303 @@
+The following example enforces the baseline Pod Security Standard on newly created
+namespaces that are not system namespaces and have no pod security admission label set.
+
+```yaml
````
(nit) This should use the shortcode for a code sample.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+For a request to be admitted, 
+it must pass all evaluations.
+
+## Clean up
```
```suggestion
## {{% heading "cleanup" %}}
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+The `spec.paramKind` field specifies the kind of resources 
+used to parameterize the policy.
+In this example,
+it is configured by `ReplicaLimit` custom resources.
```
No need for backticks here.

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+via the CEL `params` variable (e.g., `params.maxReplicas`).
+The `spec.matchConstraints` specifies what resources 
+this policy is designed to evaluate.
+Native types such as `ConfigMap` 
+can also be used as parameter references.
```
No need for backticks here.

Active voice might be better:
```suggestion
You can use custom resources for policy parameters, and you
can also use native API types (such as ConfigMap).
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+Native types such as `ConfigMap` 
+can also be used as parameter references.
+
+For each admission request, 
+the API server evaluates CEL expressions 
+of each (`policy`, `binding`, `param`) combination 
+that match the request.
+For a request to be admitted, 
+it must pass all evaluations.
```
```suggestion
For each admission request, the API server evaluates CEL expressions 
of each (_policy_, _binding_, _parameter_) combination that match the request.

For a request to be admitted, it must pass **all** evaluations. If any evaluation
rejects the request, the API server sends an error response to the client.
```

lmktfy commented on content/en/docs/reference/access-authn-authz/admission-policy-tutorial.md:
```diff
@@ -0,0 +1,303 @@
+
+Parameter resources allow a policy configuration to be separate from its definition.
+A policy can define `paramKind`,
+which outlines the Group, Version, and Kind (GVK) of the parameter resource,
```
```suggestion
which outlines the _group_, _version_, and _kind_ (GVK) of the parameter resource,
```

lmktfy reviewed (approved):
I'm minded to approve this.

Some caveats:
- MutatingAdmissionPolicy is now GA
- We don't write API types inside backticks (more of a nit)
- "For ValidatingAdmissionPolicy, ensure your cluster is version 1.30 or later." is implied (v1.30 is well outside our supported releases list)

---

But, **the perfect is the enemy of the good**. Adding this tutorial, even with snags, will improve our docs.

/approve
